### PR TITLE
Dynamic mirror round robin auto-weighting for downloads

### PR DESCRIPTION
**Pull request by cmyui**

### CODE REVIEW

File: app/repositories/beatmap_mirror_requests.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        """,
+        {"mirror_name": mirror_name},
+    )
+    if failure_rate is None:
+        return 1
+
+    # https://www.desmos.com/calculator/0am8xnwxyo
+    latency_weight = 1000 / math.log(p90_success_ms_latency)
+    failure_weight = math.exp(-10 * failure_rate)
+    weight = max(1, int(latency_weight * failure_weight))
```
Comment (cmyui) [inline app/repositories/beatmap_mirror_requests.py:52]: Goals: weight failure more heavily than latency; treat 1000ms as "quite good" for latency

Comment (cmyui) [reply app/repositories/beatmap_mirror_requests.py:52]: The difference between 0.1 and 0.01 error rate has a pretty significant effect

File: app/scheduling.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        for beatmap_mirror in self.mirrors:
+            beatmap_mirror.weight = await beatmap_mirror_requests.get_mirror_weight(
+                beatmap_mirror.name,
+            )
+            print(
+                "Beatmap mirror",
+                beatmap_mirror.name,
+                "has weight",
+                beatmap_mirror.weight,
+            )
```
Comment (cmyui) [inline app/scheduling.py]:
```suggestion
```

File: app/repositories/beatmap_mirror_requests.py (hunk 1 above)
Comment (tsunyoku) [reply app/repositories/beatmap_mirror_requests.py:52]: DESMOS ENJOYERS

**Comment by cmyui** [comment]
Working very nicely:

image

Graph available here: https://app.datadoghq.com/logs?saved-view-id=2773266

**Pull request merged**